rtrlib/aspa/aspa.c

```diff
@@ -239,16 +240,15 @@ static enum aspa_status aspa_table_update_internal(struct aspa_table *aspa_table
 
 #ifdef ASPA_UPDATE_IN_PLACE
 		if (revert && current->index == (*failed_operation)->index)
-			break;
+			// all operations up to (*failed_operation)->index will have been undone
+			return ASPA_SUCCESS;
 #endif
 
-		struct aspa_update_operation *next = (i < len - 1) ? &operations[i + 1] : NULL;
-
 #ifndef ASPA_UPDATE_IN_PLACE
-		if (array) {
+		if (array && array->data) {
 #endif
 			while (existing_i < array->size &&
-			       aspa_array_get_record(array, existing_i)->customer_asn < current->record.customer_asn) {
+			       array->data[existing_i]->customer_asn < current->record.customer_asn) {
```

Comment:
We should really use ˋaspa_get_recordˋ and ˋaspa_set_recordˋ as an abstraction layer over the data field.

Reply:
I don't see the point of using an abstraction layer for abstraction's sake. The aspa_array_get_record function does unnecessary bounds checks which I want to skip to improve performance